jgrapht-core/src/main/java/org/jgrapht/alg/color/ColorRefinementAlgorithm.java

```diff
@@ -68,279 +78,181 @@ public ColorRefinementAlgorithm(Graph<V, E> graph, Coloring<V> alpha) {
-             * init color classes with colorDegree(v) >= 1
-             * the maximum color degree is already initialised with 0
-             */
-            C.put(c, new ArrayList<>());
-            A.put(c, new ArrayList<>());
-        }
-        for(V v : graph.vertexSet()) {
-            /*
-             * init the color classes corresponding to the given coloring alpha
-             * init the color degree for every vertex with 0
-             * assign a color to every vertex (for initialization)
-             */
-            C.get(alpha.getColors().get(v)).add(v);
-            colorDegree.put(v, 0);
-            coloring.put(v, alpha.getColors().get(v));
-        }
-        // get an ascendingly sorted stack of all colors that are predefined by alpha
+    public Coloring<V> getColoring()
+    {
+        // initialize internal representation
+        ColoringRepresentation rep = new ColoringRepresentation(graph, alpha);
+
+        // get a sorted (ascending) stack of all colors that are predefined by alpha
         Deque<Integer> refineStack = getSortedStack(alpha);
-        // list of all colors that have at least one vertex with colorDegree >= 1
-        ArrayList<Integer> adjacentColors = new ArrayList<>(n);
-
-        while(!refineStack.isEmpty()) {
-            Integer r = refineStack.pop(); // analyze the next color
-
-            //calculate number of neighbours of v of color r, and calculate maximum and minimum degree of all colors
-            calculateColorDegree(r, coloring, C, A, adjacentColors, maxColorDegree, minColorDegree, colorDegree);
-
-            //calculate new partition of the colors and update the color classes correspondingly
-            calculateColorPartition(coloring, C, A, refineStack, adjacentColors, maxColorDegree, minColorDegree, colorDegree);
-
-            //reset attributes for new iteration such that the invariants are still correct
-            Iterator<Integer> adjColorIterator = adjacentColors.iterator();
-            
-            while(adjColorIterator.hasNext()) {
-                Integer c = adjColorIterator.next();
-                
-                for(V v : A.get(c)) {
-                    colorDegree.put(v, 0);
-                }
-                maxColorDegree[c] = 0;
-                A.put(c, new ArrayList<>());
-                adjColorIterator.remove();
-            }
+
+        // main iteration
+        while (!refineStack.isEmpty()) {
+            Integer currentColor = refineStack.pop();
+
+            Set<Integer> adjacentColors = calculateColorDegrees(currentColor, rep);
+
+            List<Integer> colorsToBeSplit = adjacentColors
+                .stream().filter(c -> rep.minColorDegree[c] < rep.maxColorDegree[c])
+                .collect(Collectors.toCollection(ArrayList::new));
+            colorsToBeSplit.sort(Comparator.comparingInt(o -> o)); // canonical order
+            colorsToBeSplit.forEach(color -> splitUpColor(color, refineStack, rep));
```

Comment:
I believe it's possible to merge the `.sort()` call here into the stream above using `.sorted(Comparator.comparingInt(i -> i))`.

And likewise, I believe it would then be possible to merge the call to `forEach()` into the stream as well, eliminating the need to `.collect()` into a list explicitly.

The advantage of doing it this way is that things would read a bit more idiomatically.

The disadvantage though is it would be trickier to debug, since everything would be in a stream and, by my understanding, Java debuggers aren't good yet at stepping through streams in a useful way.